model/src/main/java/org/projectnessie/api/params/CommitLogParams.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.projectnessie.api.params;
+
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.StringJoiner;
+
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+import org.eclipse.microprofile.openapi.annotations.media.ExampleObject;
+import org.eclipse.microprofile.openapi.annotations.parameters.Parameter;
+import org.projectnessie.model.Validation;
+
+public class CommitLogParams {
+
+  @NotNull
+  @Pattern(regexp = Validation.REF_NAME_OR_HASH_REGEX, message = Validation.REF_NAME_OR_HASH_MESSAGE)
+  @Parameter(description = "ref to show log from", examples = {@ExampleObject(ref = "ref")})
+  @PathParam("ref")
+  private String ref;
+
+  @Parameter(description = "maximum number of commit-log entries to return, just a hint for the server")
+  @QueryParam("max")
+  private Integer maxRecords;
+
+  @Parameter(description = "pagination continuation token, as returned in the previous LogResponse.token")
+  @QueryParam("pageToken")
+  private String pageToken;
+
+  @Parameter(description = "List of authors to filter by. The author is the original committer. "
+      + "No filtering by author will happen if this is set to null/empty")
+  @QueryParam("authors")
+  private List<String> authors;
+
+  @Parameter(description = "List of committers to filter by. This is the logged in user/account who performed the commit. "
+      + "No filtering by committer will happen if this is set to null/empty")
+  @QueryParam("committers")
+  private List<String> committers;
+
+  @Parameter(description = "Only include commits newer than the specified date in ISO-8601 format. "
+      + "No filtering will happen if this is set to null", examples = {@ExampleObject(ref = "javaInstant")})
+  @QueryParam("after")
+  private Instant after;
+
+  @Parameter(description = "Only include commits older than the specified date in ISO-8601 format. "
+      + "No filtering will happen if this is set to null", examples = {@ExampleObject(ref = "javaInstant")})
+  @QueryParam("before")
+  private Instant before;
+
+  public CommitLogParams() {
+  }
+
+  private CommitLogParams(String ref, Integer maxRecords, String pageToken, List<String> authors, List<String> committers, Instant after,
+      Instant before) {
+    this.ref = ref;
```

Comment:
Why is ref part of the commit log Params? In my mind its part of the URL and is a key component of the call. all the other params in commit log params are optional/defaultable.

Reply:
the idea was that `CommitLogParams` just wraps all the params that exist in order to **query** & **filter** the commit log and that would include everything that is required or optional. The validation process of all parameters is still happening when the endpoint is being called. 
Or were you thinking that we might potentially forget to set `ref` via the `Builder` for some tests because we don't check if `ref` ends up being non-null? I guess for that case we could add some checking logic in the builder.

Reply:
I think it makes sense to have some validation logic in the `Builder` itself, which I just added with some tests

Reply:
Im thinking of this from an API user standpoint. If the ref is in the object and I don't want any filtering then I am doing

``` java
CommitLogParams params = new CommitLogParams("ref");
handleResults(api.getLog(params));
```

Compared to:

``` java
handleResults(api.getLog("ref", null)); //or I add a helper for the no filtering case and drop the null argument
```

To me its clearer and more direct this way for a user. The `ref` is very different semantically from the optional args. @laurentgo  and @snazy you feel strongly either way?

PS I like the valildation logic though :-)

Reply:
I'm with you and I totally get your point. I had the same thoughts at the beginning, meaning I was hesitant whether `ref` should go into `CommitLogParams` or not. 
At the end I decided it's probably better to have `ref` in `CommitLogParams`, since `CommitLogParams` isn't just for filtering, but rather for all params.
Therefore `handleResults(api.getLog(CommitLogParams.builder().ref("ref").build()))` seemed to be the better choice.

I guess we could also just add a helper where we can call `api.getLog("ref")`, which builds the `CommitLogParams` object and forwards the call to `getLog(CommitLogParams)`.

Thoughts on this?

Reply:
> At the end I decided it's probably better to have `ref` in `CommitLogParams`, since `CommitLogParams` isn't just for filtering, but rather for all params.

Not sure I understand what you mean. Maybe Im just not thinking all the way through though. All the others are for filtering right?

> I guess we could also just add a helper where we can call `api.getLog("ref")`, which builds the `CommitLogParams` object and forwards the call to `getLog(CommitLogParams)`.

I would rather keep it consistent. Either `getLog("ref")` and `getLog("ref", commitLogParams)` or `getLog(comitLogParams)`

Reply:
What I mean is that `CommitLogParams` just wraps all the arguments to fetch the commit log and that includes mandatory and optional parameters. Does that make sense?

Just assume we have some endpoint `getXyz(...)` where we already have 2-3 mandatory params. And now we add a few more fields (mandatory & optional). We would have to adjust the respective method parameters everywhere where this method is being called/used. Using a wrapper class for **all** of those parameters seems like the better choice imo, as otherwise you would have a mix of mandatory parameters on the respective `getXyz(...)` method that would need adjusting, but also the optional params in the `XyzParams` class.

I don't feel strongly about having `ref` in `CommitLogParams`, so let's see what other people think.

Reply:
I moved `ref` out of `CommitLogParams` back to the original `getCommitLog(...)` method. This way, when we add/remove mandatory params, clients will know immediately by getting a compile error (rather than a runtime error when such a mandatory param is in a wrapper class)